bsp/src/org/jetbrains/bsp/project/test/environment/BspJvmEnvironment.scala

```diff
@@ -0,0 +1,256 @@
+import ch.epfl.scala.bsp4j.JvmTestEnvironmentParams
+import ch.epfl.scala.bsp4j.JvmTestEnvironmentResult
+import ch.epfl.scala.bsp4j.SourceItem
+import ch.epfl.scala.bsp4j.SourcesItem
+import ch.epfl.scala.bsp4j.SourcesParams
+import ch.epfl.scala.bsp4j.SourcesResult
+import com.intellij.execution.configurations.ModuleBasedConfiguration
+import com.intellij.execution.configurations.RunConfiguration
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.externalSystem.util.ExternalSystemApiUtil
+import com.intellij.openapi.module.Module
+import com.intellij.openapi.project.Project
+import com.intellij.psi.JavaPsiFacade
+import com.intellij.psi.PsiFile
+import com.intellij.psi.search.GlobalSearchScope
+import org.jetbrains.bsp.BspBundle
+import org.jetbrains.bsp.BspUtil._
+import org.jetbrains.bsp.data.BspMetadata
+import org.jetbrains.bsp.protocol.BspCommunication
+import org.jetbrains.bsp.protocol.BspJob
+import org.jetbrains.bsp.protocol.session.BspSession.BspServer
+import org.jetbrains.bsp.protocol.session.BspSession.BspSessionTask
+import org.jetbrains.plugins.scala.build.BuildMessages
+import org.jetbrains.plugins.scala.build.BuildToolWindowReporter
+import org.jetbrains.plugins.scala.build.BuildToolWindowReporter.CancelBuildAction
+import org.jetbrains.plugins.scala.extensions.inReadAction
+import org.jetbrains.plugins.scala.extensions.invokeAndWait
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Promise
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+object BspJvmEnvironment {
+
+  trait BspTargetIdHolder {
+    def currentValue: Option[BuildTargetIdentifier]
+    def update(value: BuildTargetIdentifier)
+  }
+
+  case class Error(msg: String) extends RuntimeException(msg)
+
+  type Result[A] = Either[Error, A]
+
+  def getBspTargets(module: Module): Result[Seq[BuildTargetIdentifier]] = {
+    BspMetadata.get(module.getProject, module)
+      .left.map(err => Error(err.msg))
+      .map(data => data.targetIds.asScala.map(id => new BuildTargetIdentifier(id.toString)))
+  }
+
+  def promptUserToSelectBspTarget(
+    project: Project,
+    targetIds: Seq[BuildTargetIdentifier],
+    holder: BspTargetIdHolder
+  ): Option[BuildTargetIdentifier] = {
+    val selected = invokeAndWait(BspSelectTargetDialog.promptForBspTarget(project, targetIds, holder.currentValue))
+    selected.foreach(holder.update)
+    selected
+  }
+
+  def promptUserToSelectBspTargetForWorksheet(module: Module): Unit = {
+    val potentialTargets = getBspTargets(module)
+    potentialTargets.foreach { targetIds =>
+      val holder = persistentHolderForWorksheet(module)
+      promptUserToSelectBspTarget(module.getProject, targetIds, holder)
+    }
+  }
+
+  def resolveForWorksheet(module: Module): Result[JvmEnvironment] = {
+    val holder = persistentHolderForWorksheet(module)
+    for {
+      workspace <- workspaceUri(module)
+      potentialTargets <- getBspTargets(module)
```

Comment:
Maybe lets filter only JVM targets, cause only they can be properly used in worksheets anyway.
This way we will not have to choose when it's not necessary:
![image](https://user-images.githubusercontent.com/3989292/80206164-8ec64000-8634-11ea-94f6-73bd9272a6f6.png)

Reply:
So, after leaving just JVM targets there will be practically available only main/test targets choice, am I right?
Do you have a use-case when you use test target in worksheets?

Reply:
I will see if I actually can filter it. The thing with targets is that they can be defined even per each source. So you may have 10 scala files in one package, which in intellij will be mapped to one module, but each of this file can be a different bsp target and have different classpath. 
You can find example of such project here:

https://github.com/olafurpg/bsp-pants/blob/scratchfile/scratchfile/src/test/scala/foobar/BUILD

Reply:
ok, I see...